libraries/chain/hotstuff/hotstuff.cpp

```diff
@@ -20,17 +20,19 @@ inline std::vector<uint32_t> bitset_to_vector(const hs_bitset& bs) {
    return r;
 }
 
-vote_status pending_quorum_certificate::votes_t::add_vote(std::span<const uint8_t> proposal_digest, size_t index,
-                                                          const bls_public_key& pubkey, const bls_signature& new_sig) {
-   if (_bitset[index]) {
-      return vote_status::duplicate; // shouldn't be already present
+bool pending_quorum_certificate::is_duplicate_no_lock(bool strong, size_t index) const {
+   if (strong) {
+      return _strong_votes._bitset[index];
    }
-   if (!fc::crypto::blslib::verify(pubkey, proposal_digest, new_sig)) {
-      wlog( "signature from finalizer ${i} cannot be verified", ("i", index) );
-      return vote_status::invalid_signature;
+   return _weak_votes._bitset[index];
+}
+
+vote_status pending_quorum_certificate::votes_t::add_vote(size_t index, const bls_signature& sig) {
+   if (_bitset[index]) { // check here as could have come in while unlocked
```

Comment:
This is not needed as you have already checked duplication.

Reply:
The comment explains why it is needed.

Reply:
In which case `add_vote` come in while unlocked? I did not find it.

Reply:
```
Thread 1:  pending_quorum_certificate::add_vote(key1)
Thread 1:     checks has_voted_no_lock() return false
Thread 1:     unlocks mutex
Thread 2:  pending_quorum_certificate::add_vote(key1)
Thread 2:     checks has_voted_no_lock() return false
Thread 2:     unlocks mutex
Thread 1:  locks mutex, calls add_strong_vote(key1), unlocks mutex
Thread 2:  locks mutex, calls add_strong_vote(key1), needs to double check that it has not already voted.
```

Reply:
This is an example of https://en.wikipedia.org/wiki/Double-checked_locking . Incidentally, I learned the pattern from Doug Schmidt which is my only claim to fame, lol. (see first reference in the wikipedia article.